### PR TITLE
ENH+FIX `to_conv1d()` and energy bugfix in `MuReNNDirect`

### LINKED ISSUE #45: `MuReNNDirect.to_conv1d()`
I suggest we have a method in `MuReNNDirect` that would compute the single-resolution equivalent impulse response of the MuReNN layer. This would be helpful for visualization in Fourier domain, for receptive fields, and for comparing computational costs.

The `to_conv1d` method should return a complex-valued PyTorch tensor of shape `(Q*J, (2**J)*T)`. It would involve IDTCWT. This can be done efficiently by introducing an object `@property` which would be computed on demand.

```
   DTCWT        conv1d        IDTCWT
δ -------> ψ_j --------> w_j -------> y_j
```

I volunteer to do this.

### LINKED ISSUE #47: normalization issue for `DTCWTDirect`
the coefficients don't seem to be normalized correctly when setting `alternate_gh=True`:
```python
N = 44100
x = torch.zeros(1, 1, N)
x[0, 0, N//2] = 1
J = 8

dtcwt = murenn.DTCWT(J=J, alternate_gh=False)
idtcwt = murenn.DTCWTInverse(J=J, alternate_gh=False)

x_phi, x_psis = dtcwt(x)
y_j = []

for j in range(J):
    y_phi = x_phi * 0
    y_psis = [x_psis[k] * (j==k) for k in range(J)]
    y_j.append(idtcwt(y_phi, y_psis).squeeze())

lp_psis = [torch.zeros(x_psis[k].shape) + 1j*torch.zeros(x_psis[k].shape) for k in range(J)]
y_lp = idtcwt(x_phi, lp_psis).squeeze()

plt.figure(figsize=(10, 3))
for j in range(J):
    y_jhat = torch.fft.fft(y_j[j])
    plt.semilogx(range(N), torch.abs(y_jhat), label=f'j={j+1}')
y_lphat = torch.fft.fft(y_lp)
plt.semilogx(range(N), torch.abs(y_lphat), label=f'lp')
plt.grid(linestyle='--', alpha=0.5)
plt.xlim(0.5, N//2)
plt.title('Frequency responses')
plt.legend()
``` 
output: 
image

### LINKED ISSUE #52: Energy issue in `MuReNNDirect`
1. to make the output signal's energy approximately at the same level crossing scales
2. a huge amount of negative valued signal appeared after `down_j`

see code below
```python
J = 8
Q = 1
T = 10
murenn_layer = MuReNNDirect(
    J=J, 
    Q=Q, 
    T=T, 
    in_channels=1,
)
# shape (B, C, Q, J, T)
Y = murenn_layer(X).detach()

# permute to (C, Q, J, T, B)
Y = Y.permute((1, 2, 3, 4, 0))

# reshape to (C*Q*J, T*B)
Y  = Y.reshape((Q*J, -1))
plt.boxplot(Y)
``` 
result
image

### PR DESCRIPTION
**Pull request by xir4n**
closes #45, resolves #47 , resolves #52 

Update:
- fix bug: normalization behavior inside `dtcwt.transform1d`
- new feature: MuReNNDirect.to_conv1d
- fix bug: normalization setting in `MuReNNDirect`

### CODE REVIEW

**Review by lostanlen — CHANGES_REQUESTED** [review]
Thank you but this implementation returns an array. I would prefer if `to_conv1d()` could return a fully functional `Conv1D` module whose `weight.data` contains the array you have. The padding should be the same.

Then in `tests/` you can check that `conv1d(x)` is almost equal to `murenn(x)`

OK?

**Comment by xir4n** [comment]
ok, I think I've got what you mean

**Comment by lostanlen** [comment]
Following our conversation today, this is both ENH and FIX now. Let me know when it's ready so I can review.

**Comment by xir4n** [comment]
What do you think I should do? Can I specify the `numpy` version in the continuous integration script to make this PR pass the checks?

**Comment by lostanlen** [comment]
tests are passing on this environment

```
(murenn) ➜  murenn git:(fafe691) conda list
# packages in environment at /Users/user/miniconda3/envs/murenn:
#
# Name                    Version                   Build  Channel
bzip2                     1.0.8                h6c40b1e_6
ca-certificates           2024.3.11            hecd8cb5_0
dtcwt                     0.14.0                   pypi_0    pypi
exceptiongroup            1.2.1                    pypi_0    pypi
filelock                  3.15.3                   pypi_0    pypi
fsspec                    2024.6.0                 pypi_0    pypi
iniconfig                 2.0.0                    pypi_0    pypi
jinja2                    3.1.4                    pypi_0    pypi
libffi                    3.4.4                hecd8cb5_1
markupsafe                2.1.5                    pypi_0    pypi
mpmath                    1.3.0                    pypi_0    pypi
murenn                    0.3.dev0                  dev_0    <develop>
ncurses                   6.4                  hcec6c5f_0
networkx                  3.3                      pypi_0    pypi
numpy                     1.26.4                   pypi_0    pypi
openssl                   3.0.14               h46256e1_0
packaging                 24.1                     pypi_0    pypi
pip                       24.0                     pypi_0    pypi
pluggy                    1.5.0                    pypi_0    pypi
pytest                    8.2.2                    pypi_0    pypi
python                    3.10.14              h5ee71fb_1
readline                  8.2                  hca72f7f_0
scipy                     1.13.1                   pypi_0    pypi
setuptools                69.5.1                   pypi_0    pypi
six                       1.16.0                   pypi_0    pypi
sqlite                    3.45.3               h6c40b1e_0
sympy                     1.12.1                   pypi_0    pypi
tk                        8.6.14               h4d00af3_0
tomli                     2.0.1                    pypi_0    pypi
torch                     2.2.2                    pypi_0    pypi
typing-extensions         4.12.2                   pypi_0    pypi
tzdata                    2024a                h04d1e81_0
wheel                     0.43.0                   pypi_0    pypi
xz                        5.4.6                h6c40b1e_1
zlib                      1.2.13               h4b97444_1
```

**Comment by lostanlen** [comment]
```
Collecting numpy<2.0.0 (from dtcwt>=0.13.0->murenn==0.3.dev0)
  Downloading numpy-1.26.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (61 kB)
```

😻

File: murenn/dtcwt/nn.py [hunk 1]
```diff
@@ -69,6 +72,84 @@ def forward(self, x):
...
+        # Create an empty list for {w_jq}
+        ws = []
+        for j in range(J):
+            # Wpsi_jr = Re[psi_j] * w_jq
+            Wpsi_jr = self.conv1d[j](psis[j].real)
+            # W_ji = Im[psi_j] * w_jq
+            Wpsi_ji = self.conv1d[j](psis[j].imag)
+            # Set the coefficients besides this scale to zero
+            Wpsis_jr = [Wpsi_jr * (1 + 0j) if k == j else psis[k].new_zeros(size=psis[k].shape).repeat(1, self.Q, 1) for k in range(J)]
+            Wpsis_ji = [Wpsi_ji * (0 + 1j) if k == j else psis[k].new_zeros(size=psis[k].shape).repeat(1, self.Q, 1) for k in range(J)]
```
Comment (lostanlen) [inline murenn/dtcwt/nn.py:135]: why convert to complex? and why multiply by `1j`?

**Review by lostanlen — COMMENTED** [review]
thanks. i just have a question on the zero-ing of complex-valued coefficients

File: murenn/dtcwt/nn.py (hunk 1 above)
Comment (xir4n) [reply murenn/dtcwt/nn.py:135]: I converted them to complex because you wanted the to_conv1d method to return a complex-valued PyTorch tensor in #45.   

Besides, in dtcwt, we have:  (I've uploaded a test to show this. )

$$
\begin{align}
idtcwt(x) &= idtcwt.tree_a(x) + idtcwt.tree_b(x)\\
&= idtcwt(x.real) + idtcwt(x.imag) 
\end{align}
$$


I think in the returned `Conv1d` instance, `conv1d.weight.real+conv1d.weight.imag` is equivalent to the case if we don't convert `Wpsi_j` to complex

there is a bug: maybe i should use $x=(1-1j)*\delta$ instead of $x=(1+0j) * \delta$ to convolve with `conv1d` to visualize the filter

I can change this into a real-valued version if this is too complicated

Comment (lostanlen) [reply murenn/dtcwt/nn.py:138]: Hello, i don't think $x = (1 - 1j) \delta$ is necessary. $x = \delta$ should suffice.

My question was not a mathematical question but more like a CS question. Usually when you do
`z = torch.complex(re, im)`
`re` and `im` are both real-valued. Here this is not the case, since you have multiplied `im` by `1j`.

Comment (xir4n) [reply murenn/dtcwt/nn.py]: Thanks for your advice, I've made the change.  
I was not clear in my previous answer. In the algorithm, only the real part of a complex tensor would be passed through Tree A, and only the imaginary part would be passed through Tree B. That's why I was converging `z` into a complex-valued tensor.

**Comment by lostanlen** [comment]
closed because of too high complexity
converted into new PR's: --> #55 and #56
